lib/auth/machineid/machineidv1/spiffe_federation_syncer.go

```diff
@@ -0,0 +1,582 @@
+	eventsCh chan types.Event
+}
+
+// runWhileLocked is the core loop of the syncer that runs on a single auth
+// server.
+//
+// Its goal is to sync the contents of trust bundles from remote clusters to
+// the local cluster. It does this by creating a goroutine that manages each
+// federated cluster.
+func (s *SPIFFEFederationSyncer) runWhileLocked(ctx context.Context) error {
+	s.cfg.Logger.InfoContext(
+		ctx,
+		"Obtained lock, SPIFFEFederation syncer is starting",
+	)
+	defer func() {
+		s.cfg.Logger.InfoContext(
+			ctx, "SPIFFEFederation syncer has stopped",
+		)
+	}()
+
+	// This wg will track all active syncers. We'll wait here until we're done.
+	wg := &sync.WaitGroup{}
+	defer func() {
+		wg.Wait()
+	}()
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// Set up state management that will hold a list of all active trust domain
+	// syncers.
+	states := map[string]trustDomainSyncState{}
+	mu := &sync.Mutex{}
+	startSyncingFederation := func(trustDomain string) {
+		mu.Lock()
+		defer mu.Unlock()
+
+		// Don't start if we're already syncing this trust domain.
+		_, ok := states[trustDomain]
+		if ok {
+			return
+		}
+
+		states[trustDomain] = trustDomainSyncState{
+			eventsCh: make(chan types.Event),
+		}
+
+		wg.Add(1)
+		go func() {
+			defer func() {
+				mu.Lock()
+				delete(states, trustDomain)
+				mu.Unlock()
+				wg.Done()
+			}()
+			s.syncTrustDomainLoop(ctx, trustDomain, states[trustDomain].eventsCh)
+		}()
+	}
+
+	// Establish our watcher, we'll use this to react instantly to changes to SPIFFEFederations.
+	w, err := s.cfg.EventsWatcher.NewWatcher(ctx, types.Watch{
+		Kinds: []types.WatchKind{{
+			Kind: types.KindSPIFFEFederation,
+		}},
+	})
+	if err != nil {
+		return trace.Wrap(err, "failed to create watcher")
+	}
+	defer func() {
+		err := w.Close()
+		if err != nil {
+			s.cfg.Logger.ErrorContext(
+				ctx, "Failed to close watcher", "error", err,
+			)
+		}
+	}()
+
+	// Wait for initial "Init" event to indicate we're now receiving events.
+	select {
+	case evt := <-w.Events():
+		if evt.Type == types.OpInit {
+			break
+		}
+		return trace.BadParameter("expected init event, got %v", evt.Type)
+	case <-ctx.Done():
+		return nil
+	}
+
+	// Fetch an initial list of all federations and start syncers for them.
+	trustDomains, err := listAllTrustDomains(ctx, s.cfg.Store)
+	if err != nil {
+		return trace.Wrap(err, "initially listing trust domains")
+	}
+	for _, td := range trustDomains {
+		startSyncingFederation(td.GetMetadata().GetName())
+	}
+
+	// Now we can start reacting to events, we'll want to start/stop syncers as
+	// needed. We'll want to start a syncer for any new trust domain, and
+	// propagate events to existing syncers.
+	for {
+		select {
+		case evt := <-w.Events():
+			s.cfg.Logger.DebugContext(
+				ctx,
+				"Received event from SPIFFEFederation watcher",
+				"evt_type", evt.Type,
+			)
+			if !(evt.Type == types.OpPut || evt.Type == types.OpDelete) {
+				continue
+			}
```

Comment:
The evt.Type == types.OpDelete is ignorred somewhere [in](https://github.com/gravitational/teleport/blob/0569da11ce3d098998f571389f72d37ed85b277c/lib/auth/machineid/machineidv1/spiffe_federation_syncer.go#L349) 
```
	if evt.Type == types.OpDelete {
				log.DebugContext(
					ctx,
					"Resource has been deleted, stopping sync loop for trust domain",
				)
				// If we've been deleted, we should stop syncing.
				return
			}
```

why not just 

```
if evt.Type != types.OpPut {
  continue
}
```